pallets/evm-accounts/src/lib.rs

```diff
@@ -0,0 +1,299 @@
+        Bounded, CheckedAdd, CheckedSub, Extrinsic, Hash, MaybeSerialize, Saturating,
+        SignedExtension, StaticLookup, Zero,
+    },
+    FixedPointNumber, FixedPointOperand,
+};
+use sp_std::{
+    convert::TryInto,
+    fmt::Debug,
+    marker::{Send, Sync},
+};
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod test;
+
+mod eth;
+
+type BalanceOf<T> = <<T as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<T>>::Balance;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use crate::eth::*;
+
+    use super::*;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_transaction_payment::Config {
+        /// The overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// The overarching call type.
+        type RuntimeCall: Parameter
+            + Dispatchable<
+                RuntimeOrigin = Self::RuntimeOrigin,
+                Info = DispatchInfo,
+                PostInfo = PostDispatchInfo,
+            > + GetDispatchInfo
+            + From<frame_system::Call<Self>>
+            + IsSubType<Call<Self>>
+            + IsType<<Self as frame_system::Config>::RuntimeCall>;
+
+        /// The type of hash used for hashing the call.
+        type CallHasher: Hash;
+
+        /// The max number of substrate accounts that are linked to a given eth address.
+        type MaxLinkedAccounts: Get<u32>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub (super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub (super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Account have been linked to ethereum address
+        AccountLinked {
+            substrate: T::AccountId,
+            ethereum: EthAddress,
+        },
+        EthCallExecuted {
+            result: DispatchResult,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// The ethereum address have an existing linkage already.
+        EthAddressAlreadyLinked,
+        /// The substrate address have an existing linkage already.
+        AccountAlreadyLinked,
+        /// The provided signature is invalid
+        BadSignature,
+        /// The substrate address is not linked to the given ethereum address
+        AccountNotLinked,
+        /// User have reached the maximum number of linked accounts.
+        CannotLinkMoreAccounts,
+    }
+
+    /// Map of one Substrate account to many EVM addresses
+    #[pallet::storage]
+    #[pallet::getter(fn evm_address_to_account)]
+    pub type Accounts<T: Config> = StorageMap<
+        _,
+        Twox64Concat,
+        EthAddress,
+        BoundedVec<T::AccountId, T::MaxLinkedAccounts>,
+        ValueQuery,
+    >;
+
+    /// Map of EVM address to Substrate account
+    #[pallet::storage]
+    #[pallet::getter(fn account_to_eth_address)]
+    pub type EthAddresses<T: Config> =
```

Comment:
I don't like that for EVM you use "address" and for Substrate "account". They are the same concepts. If we use the "account" word (for Substrate), maybe then we should call the EVM "account" for consistency of concepts.

Reply:
renamed the storages